### PR TITLE
dev: add TiDB 4.0 version planning and future version planning

### PR DESCRIPTION
**Pull request by scsldb**
### What is changed, added or deleted? 
New TiDB 4.0 version planning and future version planning

### Which version does your change affect? 
dev

### Checklist 

- [ ] Add a new file to `TOC.md`
- [x] No Tab spaces anywhere 
- [ ] Leave a blank line both before and after a code block
- [ ] Keep the first level heading consistent with `title` in metadata
- [ ] Use *four* spaces for each level of indentation except that in `TOC.md`

### CODE REVIEW

File: dev/roadmap.md [hunk 1]
```diff
@@ -1,103 +1,207 @@
...
+
+### TiKV
+
+#### 功能
+
+* 集群规模最大支持 200+节点
+* 全量物理备份恢复
+* 动态 Split,Merge 热点 Region
+* 精细的内存控制
+* Raft
```
Comment (winkyao) [inline dev/roadmap.md]:
```
* Raft 优化
	+ Joint consensus
	+ 只读副本
```

File: dev/roadmap.md [hunk 2]
```diff
@@ -1,103 +1,207 @@
...
+    + 悲观锁 GA
+    + 事务内不限制语句数量
+    + 支持 10GB 事务
+
+#### 性能
+
+* 提升 Scan 性能
+* 动态扩展工作线程数量
+* 弹性扩展只读副本
+* 优化调度系统确保 QPS 不抖动
```
Comment (winkyao) [inline dev/roadmap.md]: 优化调度系统和 GC 机制确保 QPS 不抖动

File: dev/roadmap.md [hunk 3]
```diff
@@ -1,103 +1,207 @@
...
+#### 易用性
+
+* 重构日志内容
+
+### TiFlash
+
+#### 功能
+
+* 列式存储
+
```
Comment (winkyao) [inline dev/roadmap.md]: remove the blank

File: dev/roadmap.md [hunk 4]
```diff
@@ -1,103 +1,207 @@
...
+* 通过 Raft learner 从 TiKV 同步数据
+
```
Comment (winkyao) [inline dev/roadmap.md]: ditto

File: dev/roadmap.md [hunk 5]
```diff
@@ -1,103 +1,204 @@
...
+* Loader 集成到 TiDB
+* TiDB Lightning 集成到 TiDB
+
+### 性能
+
+* TiDB Lightning 支持并行导入
+* TiDB Lightning 优化导入回放路径
+
+# TiDB 未来规划
```
Comment (lilin90) [inline dev/roadmap.md]: For the docs and docs-cn repos, we only support using one 1st-level heading.

File: dev/roadmap.md [hunk 6]
```diff
@@ -1,103 +1,204 @@
...
-- [x] 集群部署工具
-- [X] 高性能数据导入工具（lightning）
-- [X] 集群备份和恢复工具（包括全量+增量备份，mydumper + drainer/reparo）
-- [X] 改进 TiDB-Binlog 架构
-- [ ] 数据在线迁移工具（Syncer 升级版）
-- [ ] 集群诊断和分析工具
+### TiDB
+
+#### 功能
+
```
Comment (lilin90) [inline dev/roadmap.md]: So you have decided not to use the `- [ ]` format to mark what is done and what is not done?

File: dev/roadmap.md [hunk 7]
```diff
@@ -1,103 +1,204 @@
...
+    + Index Join 拆分为 Index Merge Join 和 Index Hash Join
+    + Radix Hash Join
+    + Index Merge
+    + Parallel Stream Aggregate
+    + Parallel Merge Sort
+    + Parallel Merge Join
+    + Full Vectorized Expression Evaluation
+* Indexes on Expressions
+* Multi-Index Scan
+* Join,Aggregate,Sort 算子支持外部存储
```
Comment (lilin90) [inline dev/roadmap.md]:
```suggestion
* Join，Aggregate，Sort 算子支持外部存储
```

File: dev/roadmap.md [hunk 8]
```diff
@@ -1,103 +1,204 @@
...
+
+* Binlog 支持服务高可用
+
+### TiKV
+
+#### 功能
+
+* 集群规模最大支持 200+节点
+* 全量物理备份恢复
+* 动态 Split,Merge 热点 Region
```
Comment (lilin90) [inline dev/roadmap.md]:
```suggestion
* 动态 Split，Merge 热点 Region
```

File: dev/roadmap.md (hunk 6 above)
Comment (scsldb) [reply dev/roadmap.md]: yes

File: dev/roadmap.md (hunk 5 above)
Comment (scsldb) [reply dev/roadmap.md]: If it is not hierarchical, it will look very confusing

File: dev/roadmap.md [hunk 9]
```diff
@@ -1,103 +1,204 @@
...
-- [ ] 数据在线迁移工具（Syncer 升级版）
-- [ ] 集群诊断和分析工具
+### TiDB
+
+#### 功能
+
+* 支持 TiFlash 存储引擎
+* Optimizer Trace
+* Multi-Column Statistics
+* Normal CM-Sketch with TOPN Statistics
```
Comment (zz-jason) [inline dev/roadmap.md]:
```suggestion
* 为普通 CM-Sketch 支持 TopN 统计信息
```

File: dev/roadmap.md [hunk 10]
```diff
@@ -1,103 +1,204 @@
...
-- [X] 改进 TiDB-Binlog 架构
-- [ ] 数据在线迁移工具（Syncer 升级版）
-- [ ] 集群诊断和分析工具
+### TiDB
+
+#### 功能
+
+* 支持 TiFlash 存储引擎
+* Optimizer Trace
+* Multi-Column Statistics
```
Comment (zz-jason) [inline dev/roadmap.md]:
```suggestion
* 支持多列统计信息
```

File: dev/roadmap.md [hunk 11]
```diff
@@ -1,103 +1,204 @@
...
+* Normal CM-Sketch with TOPN Statistics
+* 完善 Plan Cache 功能
+* 简单的自适应 SQL 引擎
```
Comment (zz-jason) [inline dev/roadmap.md]:
```suggestion
* 自适应 SQL 引擎
```

File: dev/roadmap.md (hunk 5 above)
Comment (lilin90) [reply dev/roadmap.md]: OK. We can put Roadmap to the website footer and remove it from docs.

File: dev/roadmap.md [hunk 12]
```diff
@@ -1,103 +1,204 @@
...
+#### 功能
+
+* 支持 batch write
+* 接入 TiFlash
+
+## Data Migration
+
+### 功能
+
+* 完善前向检查
```
Comment (GregoryIan) [inline dev/roadmap.md]:
```suggestion

```

File: dev/roadmap.md [hunk 13]
```diff
@@ -1,103 +1,204 @@
...
+* 可视化管理同步规则
+* 可视化管理同步任务
+
+### 易用性
+
+* Ctrl 工具集成数据一致性校验功能
```
Comment (GregoryIan) [inline dev/roadmap.md]:
```suggestion
```

File: dev/roadmap.md [hunk 14]
```diff
@@ -1,103 +1,204 @@
...
+* 支持 batch write
+* 接入 TiFlash
+
+## Data Migration
+
+### 功能
+
+* 完善前向检查
+* 可视化管理同步规则
+* 可视化管理同步任务
```
Comment (GregoryIan) [inline dev/roadmap.md]: add `数据同步在线校验`

File: dev/roadmap.md [hunk 15]
```diff
@@ -1,103 +1,204 @@
...
+
+### 易用性
+
+* Ctrl 工具集成数据一致性校验功能
+* Ctrl 工具输出任务详细信息
```
Comment (GregoryIan) [inline dev/roadmap.md]: not clear
```suggestion
```

File: dev/roadmap.md [hunk 16]
```diff
@@ -1,103 +1,204 @@
...
+
+#### 性能
+
+* 提升物理备份性能
+* 提升增量物理备份性能
+* 1PC
+* 支持 Storage Class Memory 硬件
+* 新 Raft 引擎
+
+## Data Migration
```
Comment (GregoryIan) [inline dev/roadmap.md]: anothe data migration？

File: dev/roadmap.md [hunk 17]
```diff
@@ -1,103 +1,204 @@
...
+
+### 功能
+
+* 下游支持 Oracle
```
Comment (GregoryIan) [inline dev/roadmap.md]: not in plan now

File: dev/roadmap.md [hunk 18]
```diff
@@ -1,103 +1,204 @@
...
+#### 高可用
+
+* Binlog 支持服务高可用
+
+### TiKV
+
+#### 功能
+
+* 集群规模最大支持 200+节点
+* 全量物理备份恢复
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: Not "物理备份", but "快速备份"

File: dev/roadmap.md [hunk 19]
```diff
@@ -1,103 +1,204 @@
...
+* 动态 Split，Merge 热点 Region
+* 精细的内存控制
+* Raft
+    + Joint consensus
+    + 只读副本
+* 事务
+    + 悲观锁 GA
+    + 事务内不限制语句数量
+    + 支持 10GB 事务
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: Better to move to TiDB?

File: dev/roadmap.md [hunk 20]
```diff
@@ -1,103 +1,204 @@
...
+
+* Binlog 支持数据高可靠
+
+### TiKV
+
+#### 功能
+
+* 增量物理备份恢复
+* 闪回到任意时间点
+* 分级存储
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: What is this?

File: dev/roadmap.md [hunk 21]
```diff
@@ -1,103 +1,204 @@
...
+* Raft
+    + 链式复制数据
+    + Witness Role
+* 存储引擎
+    + RocksDB Gaurd，compaction 按特定逻辑切分 SST
+    + 冷热数据分离
+
+#### 性能
+
+* 提升物理备份性能
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: Not "物理备份"

File: dev/roadmap.md [hunk 22]
```diff
@@ -1,103 +1,204 @@
...
+* 提升增量物理备份性能
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: Ditto

**Comment by winkyao** [comment]
Please resolve comments @scsldb

File: dev/roadmap.md (hunk 20 above)
Comment (scsldb) [reply dev/roadmap.md]: Hot and cold data separation

**Comment by lilin90** [comment]
@zz-jason @zhangjinpeng1987 @GregoryIan PTAL again~

File: dev/roadmap.md [hunk 23]
```diff
@@ -1,103 +1,203 @@
...
+
+#### 高可靠用
+
+* Binlog 支持数据高可靠
+
+### TiKV
+
+#### 功能
+
+* 增量物理备份恢复
```
Comment (zhangjinpeng1987) [inline dev/roadmap.md]: Not "物理备份" but "快速备份"

File: dev/roadmap.md [hunk 24]
```diff
@@ -1,103 +1,203 @@
...
+
+### 功能
+
+* Loader 集成到 TiDB
+* TiDB Lightning 集成到 TiDB
+
+### 性能
+
+* TiDB Lightning 支持并行导入
+* TiDB Lightning 优化导入回放路径
```
Comment (GregoryIan) [inline dev/roadmap.md]: I don't know what's meaning of it

File: dev/roadmap.md [hunk 25]
```diff
@@ -1,103 +1,191 @@
...
+## TiDB Toolkit
+
+### 功能
+
+* Loader 集成到 TiDB
+* TiDB Lightning 集成到 TiDB
+
+### 性能
+
+* TiDB Lightning 支持并行导入
```
Comment (GregoryIan) [inline dev/roadmap.md]:
```suggestion
* TiDB Lightning 支持多 lightning + importer 实例并行导入
```

File: dev/roadmap.md [hunk 26]
```diff
@@ -1,103 +1,191 @@
...
+* TiDB Lightning 优化导入回放路径
```
Comment (GregoryIan) [inline dev/roadmap.md]:
```suggestion

```

**Comment by winkyao** [comment]
@lilin90 PTAL

**Comment by lilin90** [comment]
@scsldb A display issue occurs on the PingCAP website. It does not support displaying the 2nd first-level heading on the right navigation. I suggest using only one first-level heading. Otherwise, the right navigation is not logic.

**Pull request merged**